### PR TITLE
Personalization | MyVA | cleanup feature flags mocks for tests

### PR DESCRIPTION
**Pull request by DanielMiller-agile6**
## Summary

- cleans up feature toggle mocks for tests under /personalization.  primarily used by MyVA, Profile uses it's own version.
- Modeled this after Adam's work on profile's mock, but instead of using a snake-case string converter, this uses the master list owned by platform team to look up the server names. We had a few feature toggles that didn't map character-for-character between client and server names, so we couldn't easily use a snake-case converter.
- exported a few new helpers from our feature-toggles mocking file, for any situations where more fine control over feature toggles may be needed for any tests
- attempting to generate feature toggles using for a mock object using our helper will now throw an error if the flag doesn't exist in platform master list.
- this solution also allows us to automatically load up feature toggles for our mocks by a namespace prefix, potentially removing the need to reference the toggle names anywhere other than the platform master list file, and any tests that need to manually change the toggle's value.
- Which team do you work for: **MyVa Team**

## Related issue(s)

- [Issue #65042](https://github.com/department-of-veterans-affairs/va.gov-team/issues/65042)
- https://github.com/department-of-veterans-affairs/va.gov-team/issues/72329

## Testing done

- No major functional changes, this is mainly a refactor of our existing feature-toggles mock file. Consumer modules function as expected.
- Verify that if a user tries to use a flag name that doesn't exist in master list, throws error.
- Unit and E2E tests that used the old mock all pass with the new one
- Manual testing of various scenarios with different sets of feature toggles
- tested using stringified version of object from previous commit, and compared to stringified json module post-commit (e.g. `JSON.stringify(pre) === JSON.stringify(post) // -> true`)

## What areas of the site does it impact?

These changes only affect unit and e2e testts, along with any local mocking server functionality that uses our feature toggles mock.  This is isolated to /personalization app, and is only used by MyVa and Notification Center. Profile is unaffected, since it uses alternate mock.

### Quality Assurance & Testing

- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x] Linting warnings have been addressed
- [ ] Documentation has been updated (*if necessary)

### Error Handling

- [x] Browser console contains no warnings or errors.

### Authentication

- [x] Did you login to a local build and verify all authenticated routes work as expected with a test user

## Requested Feedback
 Any feedback welcome :). This started as an experiment to see if we could do something similar to Adam's work in Profile, but also using the platform toggles list as the source of truth (instead of using a snake_case converter).  Downsides could be that this config is looking a little less straight forward and more complex.  It also doesn't really solve the question of "do we want all toggles on by default? or off?"... but it is fairly configurable (and could be made more configurable) to allow folks to choose their own conventions per app or file.

### CODE REVIEW

**Review by CBonade — DISMISSED** [review]

File: src/applications/personalization/common/mocks/feature-toggles/index.js
Comment (adamwhitlock1) [inline src/applications/personalization/common/mocks/feature-toggles/index.js:1]: I remember why I didn't use the export from `@department-of-veterans-affairs/platform-utilities/featureFlagNames` now...

This code actually will not work, and when you try and run the mock server you will get the error:

```SyntaxError: Cannot use import statement outside a module```

This is because you can't use `import` inside the mock serve Node scripts and instead have to use `require()`

Comment (DanielMiller-agile6) [reply src/applications/personalization/common/mocks/feature-toggles/index.js:1]: good catch, thanks! I hadn't run the mock server, only the cypress tests. odd that those passed tho 🤔 . I'll change this to use `require` and test some more with the mock server 👍

Comment (adamwhitlock1) [reply src/applications/personalization/common/mocks/feature-toggles/index.js:1]: hmm that might be because the cypress tests run in an es6 module kind of context instead of the Node runtime context.

IDK that you can actually make this work as is since Node is going to complain about that file when you try and require it... let me know though if you can figure it out.

Comment (adamwhitlock1) [reply src/applications/personalization/common/mocks/feature-toggles/index.js:1]: hmm that might be because the cypress tests run in an es6 module kind of context instead of the Node runtime context.

IDK that you can actually make this work as is since Node is going to complain about that file when you try and require it... let me know though if you can figure it out.

Comment (DanielMiller-agile6) [reply src/applications/personalization/common/mocks/feature-toggles/index.js:1]: yea, you may be right. I tested a bit with mocker-api, and there isn't a way (afaik) to use an async `import` statement to load platform's ES module that houses the name mapping object, at least within the context of the scripts loaded by the mocker-api cmd line interface locally.

Th only way around this I've found is to use `fs.readFileSync` to load the platform JS file as a string, and then parse out the object, convert to JSON string, and parse it as object.  I have an example of this working, but it i ugly and adds more complexity than I'd like :( ... I'd feel better about it if I knew it was a temp solution :) 

I think we could still use this approach if platform team managed this mapping as a .json file instead of a .js ES module.  If the master list is json, we could more easily support multiple files for different module contexts (e.g. featureFlags.js, fatureFlags.cjs, etc), which means client-side app code and mocking server code would be referencing the same exact list. Not sure if there could be other reasons to avoid doing like this, but would allow for more interoperability and avoid relying on strict naming conventions.

@allisonlu @adamwhitlock1 Maybe we could chat about this at our next pairing hour? Going to put this PR on hold for now 👍

Comment (DanielMiller-agile6) [reply src/applications/personalization/common/mocks/feature-toggles/index.js]: I updated this PR with an example of this using a json file for client-to-server flag name mappings.  switching this PR to "in progress" since I'm testing out an idea with some platform changes and I wanted to test this out a bit with CI.  This builds off of changes from earlier, but also supports running mocker-api locally.  Since this changes things at the platform level, we can use this PR to discuss this and/or other approaches for resolving the problem :)

File: src/applications/personalization/common/mocks/feature-toggles/index.js
Comment (DanielMiller-agile6) [inline src/applications/personalization/common/mocks/feature-toggles/index.js:2]: this module is passed directly to mocker-api, running in CommonJS mode, so i don't believe there's a way to use platform aliases in babel config, unless we were to pre-compile the 3rd party code when running the local mock server (which seems excessive imo). using cjs module seems a bit simpler, since it's only supporting the local dev use-case.

is there another way to get this relative path, w/o so much backtracking? since this is a cjs module, and only needed in node, we could use file and path modules here, but if there's a prefered way of doing this in common js modules for va.gov, let me know :)

File: src/platform/utilities/feature-toggles/featureFlagNames.json
Comment (DanielMiller-agile6) [inline src/platform/utilities/feature-toggles/featureFlagNames.json:1]: if PR looks ok after any more changes, I'll make sure to update to latest featureFlagNames in `main` before re-generating the JSON. I'm basically just importing the old js module and calling `JSON.stringify` on it, so it can match the latest config from main whenever we merge

File: src/applications/personalization/common/mocks/feature-toggles/index.js
Comment (adamwhitlock1) [reply src/applications/personalization/common/mocks/feature-toggles/index.js:2]: You could do:

``` js
const path = require('path');

const featureTogglePath = path.join(
  process.cwd(),
  '/src/platform/utilities/feature-toggles/featureFlagNames.cjs',
);

const FEATURE_FLAG_NAMES = require(featureTogglePath);
```

But eslint complains with this error: 'Calls to require() should use string literals eslint [import/no-dynamic-require](https://github.com/import-js/eslint-plugin-import/blob/v2.24.2/docs/rules/no-dynamic-require.md)' So i'd say leave it as it is.

**Review by CBonade — DISMISSED** [review]

**Comment by DanielMiller-agile6** [comment]
doh, resolving conflicts with the featureFlagNames.js file using a rebase might have dismissed reviews... hopefully this is the last time this changes or needs review :)

json has been updated for this branch and matches the latest version of the previous object. tested using stringified version of object from previous commit, and compared to json module post-commit.

**Review by adamwhitlock1 — DISMISSED** [review]
As a side note or task to set up once this is merged, there are some docs here: https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-5.Addthefeaturetogglenametovets-websitebyupdatingfeatureFlagNames.js

That will need to be updated. If you scroll to the bottom of the page there is a link to create a ticket for the content team to update the content around the feature flags.

Thanks for getting this to the finish line @DanielMiller-agile6 !

**Comment by DanielMiller-agile6** [comment]
k, ticket created 👍 
https://github.com/department-of-veterans-affairs/va.gov-team/issues/77938

@adamwhitlock1

**Review by CBonade — DISMISSED** [review]

**Review by CBonade — APPROVED** [review]

**Pull request merged**
